core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -339,7 +339,7 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
    * to the default one.
    */
   final def on(ec: ExecutionContext): IO[E, A] =
-    IO.shift(ec) *> self <* IO.shift
+    IO.shift(ec).bracket_(IO.shift)(self)
```

Comment:
It works, but in event of failure, it only does as a side-effect of .observe in `catchError`, when async is not completed by the time observe runs it will spawn a new task, but if it is completed the fiber will resume on the same thread. So this also shift back to the RTS even though there are no explicit async boundaries:

```scala
val ec = ExecutionContext.fromExecutorService(Executors.newCachedThreadPool())

def printThread: IO[Nothing, Unit] =
  IO.sync(println(Thread.currentThread()))

new scalaz.zio.RTS {}.unsafeRun {
  for {
    _ <- printThread
    _ <- IO.shift(ec)
    _ <- printThread
    _ <- IO.unit.bracket_(IO.sync(Thread.sleep(1000)))(IO.fail(())).catchAll(_ => IO.unit)
    _ <- printThread
  } yield ()
}
```
```
Thread[main,5,main]
Thread[pool-1-thread-1,5,main]
Thread[zio-2-1169146729,5,zio]
```